design-proposals/kubebmc.md

```diff
@@ -0,0 +1,156 @@
+
+## Non Goals
+
+- Providing BMC functionalities for bare-metal machines
+
+## Definition of Users
+
+- Users who want to do out-of-band management to the VMs within a KubeVirt environment directly using bare-metal-like methods, e.g., IPMI or Redfish
+- Users who want to install software like OpenShift or OpenStack on VMs locally within a KubeVirt environment using existing deployment tools
+
+## User Stories
+
+- As a developer, I want to deploy a multi-node OpenShift cluster using its native installer in KubeVirt VMs.
+- As a tester, I need to quickly spin up and destroy OpenStack clusters in a virtual environment for testing purposes.
+
+## Repos
+
+No existing repos under the KubeVirt organization will be impacted. KubeBMC augments the KubeVirt ecosystem as an optional add-on providing an operator and the BMC implementation.
+
+# Design
+
+KubeBMC will act as a bridge between KubeVirt VMs and the BMC functionalities, using Kubernetes services to expose these capabilities. KubeBMC consists of two components:
+
+- **kubebmc-controller**: A typical Kubernetes controller reconciles the KubeBMC, VirtualMachine, and Service objects
+- **kbmc**: A BMC simulator for serving IPMI and Redfish endpoints and translating the incoming requests to native KubeVirt API calls
+
+Each VirtualMachine will have its own KubeBMC object created by `kubebmc-controller` during the VM creation. The controller will further process the KubeBMC object to complete the BMC scaffold. This includes a BMC emulator Pod, i.e., `kbmc`, and its Service object working together to provide the BMC functionalities to various clients over the network.
+
+Inside the BMC emulator, the IPMI/Redfish requests are translated to native KubeVirt API calls and sent to the API server. It's essentially a combination of IPMI/HTTP server and Kubernetes/KubeVirt client.
+
+The following graph illustrates the entire KubeBMC workflow:
+
+```mermaid
+flowchart LR
+    controller["kubebmc-controller"]
+    cr["kubebmc CR"]
+    kbmc-pod["kbmc Pod"]
+    kbmc-svc["kbmc Service"]
+    controller-.->|watches|cr
+    cr-.->|owns|kbmc-svc
+    cr-.->|owns|kbmc-pod
+    client--->|IPMI|kbmc-svc
+    kbmc-svc-->kbmc-pod
+    kbmc-pod-->|HTTP|apiserver
+    apiserver-->|modifies|vm
+    vm-->|creates|vmi
+```
+
+It's also possible for VMs being created to opt out of such a functionality. This could be done by annotating the VirtualMachine object with a specific key-value pair.
+
+## API Examples
+
+The proposed KubeBMC CRD looks like the following:
+
+```yaml
+kind: KubeBMC
+metadata:
+  name: openshift-cluster-node-0
+spec:
+  # A Secret object reference for the IPMI/Redfish authentication.
+  # It should contain a pair of username and password.
+  authSecretRef:
```

Comment:
Do we need `Ref` in these attribute names?

Reply:
We could get rid of it :)